march_hardware/src/PDOmap.cpp

```diff
@@ -6,129 +6,151 @@
+  { IMCObjectName::QuickStopDeceleration, IMCObject(0x6085, 32) },
+  { IMCObjectName::QuickStopOption, IMCObject(0x605A, 16) }
+};
 
-void PDOmap::addObject(IMCObjectName objectname)
+void PDOmap::add_object(IMCObjectName object_name)
 {
-  if (this->allObjects.count(objectname) != 1)
+  if (PDOmap::all_objects.find(object_name) == PDOmap::all_objects.end())
   {
-    ROS_WARN("IMC object does not exist (yet), or multiple exist");
+    ROS_WARN("IMC does not exists in in initialised objects so can not be added to PDO");
     return;
   }
-  else if (this->PDOObjects.count(objectname) != 0)
+
+  if (this->PDO_objects.count(object_name) != 0)
   {
-    ROS_WARN("IMC object is already added to PDO map");
+    ROS_WARN("IMC object %i is already added to PDO map", object_name);
     return;
   }
-  else
+
+  this->PDO_objects.insert({ object_name, PDOmap::all_objects[object_name] });  // NOLINT(whitespace/braces)
+
+  int total_used_bits = 0;
+  for (std::pair<IMCObjectName, IMCObject> PDO_object : this->PDO_objects)
   {
-    this->PDOObjects[objectname] = this->allObjects[objectname];
+    total_used_bits = total_used_bits + PDO_object.second.length;
+  }
+
+  if (total_used_bits > this->nr_of_regs * this->bits_per_register)
+  {
+    ROS_FATAL("Too many objects in PDO Map (total bits %d, only %d allowed), PDO object: %i could not be added",
+              total_used_bits, this->nr_of_regs * this->bits_per_register, object_name);
+    throw std::exception();
   }
 }
 
-std::map<enum IMCObjectName, int> PDOmap::map(int slaveIndex, enum dataDirection direction)
+std::map<IMCObjectName, int> PDOmap::map(int slaveIndex, enum dataDirection direction)
 {
-  this->sortPDOObjects();
-  int reg;
-  int SMAddress;
+  this->sort_PDO_objects();
+
   if (direction == dataDirection::miso)
   {
-    reg = 0x1A00;
-    SMAddress = 0x1C13;
+    return configurePDO(slaveIndex, 0x1A00, 0x1C13);
   }
   else if (direction == dataDirection::mosi)
   {
-    reg = 0x1600;
-    SMAddress = 0x1C12;
+    return configurePDO(slaveIndex, 0x1600, 0x1C12);
   }
   else
   {
-    ROS_ERROR("Invalid dataDirection argument");
+    throw std::runtime_error("Invalid dataDirection argument in PDO mapping");
   }
-  // Clear SyncManager Object
-  sdo_bit8(slaveIndex, SMAddress, 0, 0);
-  int startReg = reg;
-  int lastFilledReg = reg;
-  int sizeleft = this->bitsPerReg;
-  int counter = 0;
-  int byteOffset = 0;
-  while (this->sortedPDOObjects.size() > 0)
+}
+
+std::map<IMCObjectName, int> PDOmap::configurePDO(int slaveIndex, int baseRegister, int baseSyncManager)
+{
+  int counter = 1;
+  int currentReg = baseRegister;
+  int sizeLeft = this->bits_per_register;
+  std::map<IMCObjectName, int> byte_offsets;
+
+  sdo_bit8(slaveIndex, baseRegister, 0, 0);
+
+  for (const auto& nextObject : sorted_PDO_objects)
   {
-    // Check if register is still empty
-    if (sizeleft == this->bitsPerReg)
+    sizeLeft -= nextObject.second.length;
+    if (sizeLeft < 0)
```

Comment:
What if sizeLeft equals 0?

Reply:
Ah yes I see. It looks a bit counterintuitive, but if you think it's ok that's fine.